### DISCUSSION TITLE
Release 4.0

### DISCUSSION THREAD

**Discussion by revoxhere** [revoxhere/duino-coin#1726] [General]
## Hello Duino miners!
Woah, it's the biggest update we've had this year probably. Most importantly you can now use the universal ESP code and benefit from the new reward system (when using the newest version). The PC and AVR Miners have also been updated to fix some of the bugs and provide more user friendly experience. Please note that I'll probably be still adjusting the earnings in the upcoming days, but I don't plan on making any major changes. Only the Arduino code didn't receive any changes, so you don't have to reflash your boards this time. And last but not least, happy new year! 🎉 🎆 
## What's new, what's changed...
* Release 4.0 pull by @revoxhere in https://github.com/revoxhere/duino-coin/pull/1723
  * All-new, more efficient and universal ESP code for **all Espressif boards**, by @matteocrippa and @revoxhere
    * Better performance (66 kH/s on ESP8266-based boards, 84 kH/s on dual-core ESP32-based boards)
    * Separate settings file (**Settings.h**) for easy configuration
    * New Duino IoT options (DS18B20 and internal temp sensor for new ESP32s)
    * Dashboard should not interfere with mining anymore
    * Better mining stability and faster node fetching
  * PC_Miner and AVR_Miner changes
    * Added a print queue to prevent logs getting messed up while mining with many threads/boards
    * Miners now display per thread and total hashrate
    * Added some missing internalization strings
    * Fixed greeting message not displaying properly with more than one board setup in AVR Miner
    * Updated share messages and log colors
    * Added missing dynamic window title for the PC Miner
    * Minor bugfixes and improvements
  * Changes in the reward system
    * Some difficulties have been readjusted and all have had their earning rates improved
    * Raspberry Pi and PC Mining earnings buffed
    * Single core ESP32 boards will now use a separate difficulty to make mining for them more forgiving
    * Difficulty calculation for PC miners now uses an updated Kolka V3 function for smoother difficulty changes
* Getting started pages updated with a workaround for the dependency install by @revoxhere in 60e95f9198a5f1e4a0c898317d2c4f68bb68ec38
* Getting started pages updated to support the new ESP code by @revoxhere in 60e95f9198a5f1e4a0c898317d2c4f68bb68ec38
* Added Traditional Chinese README by @BX2AFO in https://github.com/revoxhere/duino-coin/pull/1707
* Added Traditional Chinese link to the main README by @BX2AFO in https://github.com/revoxhere/duino-coin/pull/1710
* Updated Czech translation by @fpetru104 in https://github.com/revoxhere/duino-coin/pull/1706
* Update README_zh_TW.md by @BX2AFO in https://github.com/revoxhere/duino-coin/pull/1715
* Update README_es_LATAM.md by @Erick2317 in https://github.com/revoxhere/duino-coin/pull/1714
* Update README_zh_TW.md by @BX2AFO in https://github.com/revoxhere/duino-coin/pull/1718
* Update benchmarks and add Release 4.0 performance goals by @revoxhere in https://github.com/revoxhere/duino-coin/pull/1724
* Update project maintainer email by @Tech1k in https://github.com/revoxhere/duino-coin/pull/1713

## New Contributors - welcome and thanks for your input!
* @matteocrippa made their first contribution in https://github.com/revoxhere/duino-coin/pull/1723
* @BX2AFO made their first contribution in https://github.com/revoxhere/duino-coin/pull/1707
* @Erick2317 made their first contribution in https://github.com/revoxhere/duino-coin/pull/1714

### ⚠️ NOTICE: .exe binaries may be detected as a virus by your antivirus software. This is a **false positive** caused by pyinstallers bootloader, read about it [here](https://stackoverflow.com/questions/43777106/program-made-with-pyinstaller-now-seen-as-a-trojan-horse-by-avg). Duino-Coin is not a virus.

**Full Changelog**: https://github.com/revoxhere/duino-coin/compare/3.9...4.0
As always, please report any encountered issues. Happy mining!

<hr /><em>This discussion was created from the release <a href='https://github.com/revoxhere/duino-coin/releases/tag/4.0'>Release 4.0</a>.</em>

**Comment by CarlosMariaVazquez** [comment]
cuando actualice solo una placa de mi minero a la version 3.9, dejo de funcionar todo el minero y quedo el esp32 con la version 3.9 en pantalla , trabado sin funcionar, con el esp32 desconectado inclusive.

**Reply by Jamtart420** [reply]
Thats weird I have tried to update my miners to **[4.0](https://github.com/revoxhere/duino-coin/releases)** with no luck but **5** out **22** miners of mine are still running on **[3.5](https://github.com/revoxhere/duino-coin/releases/tag/3.5)** and the others running on **[3.9](https://github.com/revoxhere/duino-coin/releases/tag/3.9)** and not one on **[4.0](https://github.com/revoxhere/duino-coin/releases)**. just thought I would tell what my miners are doing thanks for your info.

**Comment by CarlosMariaVazquez** [comment]
![image](https://github.com/revoxhere/duino-coin/assets/51857569/45722501-f382-4ad6-8e98-77191f416b55)

**Comment by CarlosMariaVazquez** [comment]
![image](https://github.com/revoxhere/duino-coin/assets/51857569/75b3db76-341c-402e-b338-503ca3b628a2)
